### ISSUE TITLE
Missing $exceptions checks in PHPMailer.php

### ISSUE THREAD

**Issue by orlitzky** [PHPMailer/PHPMailer#1697]
I noticed the following error in our logs last night:

```
Mar 28 13:43:32 http2 apache2: PHP Fatal error:  Uncaught Error: Class 'PHPMailer\PHPMailer\Exception' not found in /usr/share/php/PHPMailer/src/PHPMailer.php:1411
Stack trace:
#0 /usr/share/php/PHPMailer/src/PHPMailer.php(1316): PHPMailer\PHPMailer\PHPMailer->preSend()
#1 /var/www/example.com/www/public/ticketmail.php(22): PHPMailer\PHPMailer\PHPMailer->send()
#2 {main}
  thrown in /usr/share/php/PHPMailer/src/PHPMailer.php on line 1411
```

The `Exception` class could not be loaded because we're not using the autoloader, and because (I thought) we weren't using exceptions. However, that line is missing its `if ($exceptions)` check,

```php
// Refuse to send an empty message unless we are specifically allowing it                                                                          
if (!$this->AllowEmpty and empty($this->Body)) {
  throw new Exception($this->lang('empty_message'), self::STOP_CRITICAL);
 }
```

The line in question is https://github.com/PHPMailer/PHPMailer/blob/master/src/PHPMailer.php#L1462

It looks to me like there are a few more instances of the same problem in that file. Rather than bombard you with line numbers, I think it suffices to search for "throw" in a text editor and see which ones are missing the check.

**Comment by Synchro** [comment]
The `$exceptions` flag means that _external_ exceptions are not thrown, but they are still used internally. I really recommend using composer and the autoloader since problems like this just go away.

**Comment by orlitzky** [comment]
Nothing would go away: this still would have crashed, because an uncaught exception was thrown.

What's the difference between an internal and an external exception? Maybe this is easy to remedy with an update to the docs. Ultimately the problem is that `send()` wasn't wrapped in a `try/catch` block, because I thought the docs were telling me it wasn't necessary.

**Comment by Synchro** [comment]
Read the code - [that exception gets caught later on](https://github.com/PHPMailer/PHPMailer/blob/master/src/PHPMailer.php#L1511), and only rethrown if exceptions are enabled. This should be the case with all internal exceptions: if they are disabled, indeed you do not need a try/catch wrapper around `send()`. That doesn't mean that you don't need the Exception class, but it does mean you don't need the try/catch block.

**Comment by orlitzky** [comment]
Ok, I see what's happening, thanks for explaining. IMO the end user shouldn't be forced to load the classes that the PHPMailer class uses internally, but we're never going to use composer so I guess I can live with it.

**Comment by Synchro** [comment]
Why don't you want to use composer?

**Comment by orlitzky** [comment]
I don't think you want to open that can of worms =)

It pretends to be a package manager, but it isn't. It implements only the fun/easy parts of what a package manager needs to do, and winds up basically being a front-end for curl and `git clone`. None of the hard parts of package management (that the other package managers have been working on for 20 years) are there. For example,

  * You can't depend on anything other than PHP packages. What if I need to shell out to a perl script? Impossible.
  * There's no build phase, so you can't e.g. install extensions that need you to run `./configure && make`.
  * There's no (pre-)install phase, so you can't strip certain things out for production. For example, everyone should run the tests before installing a package, but most packages don't even _ship_ the tests, because composer is stupid and installs them to your public document root. Oh, and the tests don't work to begin with, because they're all looking for an `autoload.php` file that isn't in the release tarball.
  * There's no secure central repository that it pulls from. As a system administrator, I have no idea what composer is going to download and run on my servers. Or who wrote it. Or if its GPG signed. Or if it's fetched over plain HTTP.
  * There's no security process if a "composer package" has a vulnerability. By this I mean that there's no analogue for something like Debian or Red Hat's security team. If there's a critical remote vulnerability in a composer package and if no one fixes it, then "composer install" will happily install it for the rest of eternity.
  * Composer doesn't install to a central location, so there's no way to do security updates on an entire server. How would I even know which directories I need to run "composer update" in? I have hundreds of websites to manage on multiple servers. This is the difference between a five-minute "emerge --update" and hours figuring out how to update all of the composer sites.
  * Since composer bundles everything, it promotes incompatibility between packages. Upstream authors who have gone full-composer break their API regularly, using "just use composer" as their excuse. But this makes it impossible for people who don't want to bundle everything to install a single copy of a library -- that single copy won't be compatible with everything.

Now, most of those problems are shared by every other language-of-the-week "package manager," like ruby's bundler, Haskell's stack, and so on. But what really makes composer the worst of them all is,

  * Composer encourages upstream packages to break their own code, by deleting the necessary `require` statements, and insisting that some magic sauce be added later to make everything just work.

With all of those other language-specific package managers, when people choose to use them -- that's fine, you're only wasting your own time. I can still choose to use the Debian/RedHat/whatever package, and that distro package solves all of the problems listed above, and isn't very hard to create. But composer is special: I can't install a composer package without composer! The packages themselves are outright broken when you download them. You need a special tool (i.e. composer) to generate a magic autoload file before anything will work, and that autoload file basically enforces all of the problems that I listed above, meaning that I can't just do things right on my own machines.

Specifically, in this case, it means that I have to start re-implementing the private details of the PHPMailer class if I want to use a single up-to-date copy in `/usr/share/php/PHPMailer` rather than 1,000 insecure outdated copies strewn all over the server. And regardless of how it gets done, having the end user load the internal implementation details of your classes is retarded and backwards. With an autoloader, it's also a hell of a lot slower. The final knife in the back is that all of this stuff just worked a few years ago. Everything was great; switching to composer broke it.

To summarize, all language-specific package managers (including composer) are bad reimplementations, by inexperienced programmers, of the package managers that already exist. Relative to those other package managers, they have only downsides. They create security vulnerabilities and maintenance nightmares by ignoring the past 30 years of software engineering and system administration experience. I simply can't use them: there's not enough time in the day to update 1,000 websites one-at-a-time. I also cannot with a clear conscience leave those 1,000 sites vulnerable to known exploits. There's a better way to do it, and that way existed long before composer (or bundler, or stack, or pip) did.

**Comment by Synchro** [comment]
I think the main thrust of your complaint is about where updates are applied. Complaining that composer doesn't update all packages on a server across multiple apps is like complaining that apt doesn't update packages across multiple servers, or that it fails to update the packages across all your docker containers. Complaining that you need to install composer first is pretty fatuous - guess what, you also have to install PHP to run PHP, and why can't I use apt on windows?

There is a long history of repackaging things like pear packages and PHPMailer as apt packages, but generally speaking, it's also been a great way of breaking everything at once, which is why that's become so unpopular. It also adds to the burden of maintainers; composer needs to be able to run across many different OSs, so any developer is going to want to create a single package that works everywhere that PHP can run, not tied to a single OS-dependent package manager.

FYI there is an exact parallel to the Debian and RH security feeds in the form of Symfony's checker, and Roave's vulnerability project, which prevent installation of packages with known vulnerabilities by creating version conflicts.

I don't think I have ever wanted the ability to update multiple apps at the same time to an identical set of dependencies, since every app can be different and have its own requirements, but if I did want to do something that complex, I'd expect to do it with tools like ansible which can do a thousand different composer installations in its sleep. This is much, much worse in python, as a version mismatch (in either direction) by the tiniest amount is usually fatal, necessitating revolting constructs like virtualenv.

**Comment by orlitzky** [comment]
My complaint isn't that I have to install composer first, it's that **I have to install the package with composer**. Otherwise, it won't work. If I install composer with my distro's package manager, and if I install PHPUnit with my distro's package manager,... nothing works, because `autoload.php` is missing, and I have no way to generate it.

I'm not sure what you're getting at with the server/docker analogy. I still have about 100 times more websites than I do web servers, so using apt to upgrade dependencies is about 100 times faster than composer would be.  Sure, if I could run apt on all of my servers at once, it would be another 20x faster... but I don't see how that invalidates the fact that switching to composer would waste an absurd amount of time relative to what I already do.

The other two reasons you give are just reasons that you (or somebody else) might not want to use distro packages. First of all, that's fine -- I don't want to tell you how to do your job, or how to spend your own free time. But the problems you perceive can be (and have been) solved a long time ago. People have been using distro packages in their modern form for about 20 years, and the major distros have something like 20,000 packages in their repositories that all play nicely together. The trade-off is that the distro maintainer has to spend a little bit of time up-front to make things work nice, but then millions of users save time because they get something that works. Composer makes it easy for the developer, and then wastes the user's time a million times over.

In any case, given that I am going to upgrade things like PHPMailer on hundreds of websites all at once, and that I already use my distro's package manager to do that for everything except certain PHP packages, introducing composer into the mix would only waste time. It's also insecure, keeps non-PHP dependencies implicit, can't depend on extensions, etc. But ultimately it would just be a waste of my time. Routine updates are not what I do for fun. Sorry for the rant, but I hope I at least answered your question.

**Comment by Synchro** [comment]
I'm saying that every one of your sites can have dependencies on different versions of the same packages. How would you propose to use a system package manager to deal with that? FWIW, you can also install packages system-wide using composer that are then available to all apps.

20k packages is small potatoes; [there are 1.5 million installable versions of 200,000 packages in packagist alone](https://packagist.org/statistics), not including private repos, and that does 550 million installations/month.

You can always do what the entire world that's on shared hosting also does - run composer locally, which gets all your dependencies and generates your autoload script. Since all it does is create a bunch of PHP files (and you already have PHP), you're then free to upload it along with all your other stuff. And it does all that in an OS-agnostic way.

You can depend on extensions. Just add a dependency on `ext-` followed by the extension name; you'll see that PHPMailer has a bunch.

There are mechanisms to allow package managers to cooperate, for example there are adaptors to let you install npm packages from composer.

If it really worries you that much, then you can always use [the Debian package of PHPMailer](https://packages.debian.org/sid/libphp-phpmailer).

**Comment by orlitzky** [comment]
> I'm saying that every one of your sites can have dependencies on different versions of the same packages. How would you propose to use a system package manager to deal with that?

This is not a problem with adult software. Pick... your favorite mail client. How many versions of OpenSSL can it compile against? Even with a compiler and type system in the way, like ten years worth of versions? If there's a problem with PHP code in this respect, it's because of one of my earlier bullet points:

* Since composer bundles everything, it promotes incompatibility between packages...

Generally, everyone will remain compatible with a few major versions of the things they use. When a new major version of a library comes out, its consumers add support for it. Once everyone has support for it, you upgrade. Nobody has to know about any of this, though, because the package manager does it all for you. Your quality of life is greatly improved by never having to think about this nonsense.


> FWIW, you can also install packages system-wide using composer that are then available to all apps.

That's a step in the right direction, but it's still a half-assed reimplementation of what I already have. At best, it's going to take me twice as much time to run my regular updates *and* composer updates.

> You can always do what the entire world that's on shared hosting also does - run composer locally, which gets all your dependencies and generates your autoload script. Since all it does is create a bunch of PHP files (and you already have PHP), you're then free to upload it along with all your other stuff.

This idea has promise, but your easy implementation doesn't work. The autoloaders that composer generates only work with relative paths, so either:

* I have to hand-write an autoloader that will work from a global system installation. This works, and is what the Debian/Fedora/Gentoo packages do, but is an insane amount of effort. We have to reverse engineer every version of every library to figure out which `require` calls should be there.
* I have to run composer a thousand times and upload a thousand autoload.php files whenever some library releases an update (just as much work as using composer directly, i.e. too much)


> You can depend on extensions. Just add a dependency on ext- followed by the extension name; you'll see that PHPMailer has a bunch.

That doesn't do anything. Composer can't *install* the extensions, it will just say "sorry" if one isn't there. A real package manager can force you to rebuild PHP itself with the e.g. the imap extension, if your package needs it. This is all old news.


> If it really worries you that much, then you can always use the Debian package of PHPMailer.

Hand-writing an autoloader for PHPMailer is easy. I created and maintain the PHPMailer package on Gentoo. But something like PHPUnit? There are thousands of classes that need to be hard-coded, because all of its dependencies need hand-written autoloaders too. We have a pull request open right now where multiple people have wasted probably fifty hours trying just _to replicate the `require` statements that upstream deleted for composer_. We still don't have a modern version of PHPUnit because of all the work involved. And keep in mind that PHPUnit worked fine before composer. Yes, with the help from the Fedora guys, we can hand-write out own autoloaders. But doing _that_ is a huge waste of time too! When we the distro maintainers can't make PHP packages work, end users are forced to use composer for their sites. Which (for the reasons I've been beating to death) wastes their time. There's a massive opportunity cost; the sysadmins of earth could be doing better things.

Some web developer did not just think up a solution to dependency management that had escaped the best UNIX admins for 50 years. Composer is way behind the times, and enshrines a number of terrible practices for a community that already has a reputation for bad code and worse security. No matter how good you make it, it will still be worse than what I already have. And if you're asking me to use it *in addition* to what I already have, then you're just asking me to waste my time. I have a tiny bit of self-respect these days and refuse to do it.

**Comment by Synchro** [comment]
I think we're seeing this from very different directions. My usual arrangement (and that of most people I encounter that use composer) is multiple servers collectively running a single app, not thousands of copies of similar apps on the same server. Without language-native, OS-independent dependency management, most modern PHP apps would be impossible. Also bear in mind that it manages more than just external dependencies.

> fifty hours trying just to replicate the require statements 

For any non-trivial app, manually organised `require` calls are a total train wreck that I would _never_ want to return to; they're utterly unreliable and massively inefficient - if you think you need `require_once`, you've gone very wrong already. You're deliberately recreating the exact problem composer was created to address. Of all the packaging systems I've used (pip, npm, apt, cpan, yum, bower, gem, etc), composer has been by far the easiest, most consistent, and most reliable of them, largely because it made a deliberate effort to avoid the problems that they ran into, while adopting the approaches that they excelled at. PHP has come an awfully long way since "the bad old days", and much of what you assert is simply no longer true, and is frankly insulting.

**Comment by orlitzky** [comment]
I think by arguing any further, I'm just letting composer indirectly waste more of my life. I don't doubt that a lot of people find composer useful, but it's wrong for a lot of things too; specifically the exact things that I need to do as part of my job and in my hobby as a distribution maintainer.

I apologize for the insult: I am obviously not a fan of the tool, but I'm trying to criticize the thing and not the people.

**Comment by martinmubaiwa** [comment]
Orlitzky, you are so sooooooo right. I have bumped into many situations where I feel I am being encourage to use composer BUT I I feel using it takes away control from me. As versatile full stack developer I work with many languages and dependencies and I so I need to know what’s happening behind the scenes during installation. So if a tool is not doing what I need, I should be able to take full control of the situation.
